### ISSUE TITLE
Feature enhancement: Taking input file paths from a file instead of passing them on command line

### ISSUE THREAD

**Issue by videoguy** [lsh123/xmlsec#461]
We use xmlsec tool as part of our build on CICD infrastructure to sign couple of thousand xml files. The way xmlsec is invoked, the input file path and output file path have to be passed on command line.
That required launching the app 2000+ times. It is taking close to 10 minutes to complete the process.
I am wondering if there are ways to make it faster.
I am thinking out loud here. How about xmlsec reads input file and optionally output file paths from a file that is passed parameter.
Something like this.

xmlsec <bunch of parameters> --inputbatch <path to file with multiple lines where each line contains one file path>

Then xmlsec can iterate over this list instead of calling program spawning it 1000s of times.

If you think of oher ways to make this faster, please share.
We use xmlsec 1.2.18 (openssl) on Windows 10.

**Comment by lsh123** [comment]
A couple thoughts:

1) Probably the easiest thing to try is to run this process in parallel. Even with a single CPU there is a lot of IO and you should see benefits in splitting the 2000 files into (say) 10 batches of 200 files each and then running these in parallel (search for "start" command on Windows).

2) I suspect that even single threaded batch file might be somewhat slow since you are using a lot of public key cryptography.

3) The batch file will need to contain both input and output filenames.

Let me think about it. This is definitely doable but might be more complicated than it looks like.

**Comment by lsh123** [comment]
Another option for you to try is #462 (already merged to master). You can do something like this:

``
xmlsec1 .... --output c:\tmp\{inputfile}.out file1.xml file2.xml file3.xml ...
``

This will produce output files c:\tmp\file1.out, c:\tmp\file2.out, c:\tmp\file3.out ... (the {inputfile} in the output file name is replaced with the input filename and multiple input files are allowed in the command line). Try to combine this with parallel execution described above --> I think this should reduce execution time significantly.

**Comment by videoguy** [comment]
I want to give this change a try.
How do I get access to prebuilt binaries with these changes?

One thing you might want to keep in mind is the max buffer length limitations when you try to launch a command with many file names on Windows. The command could get truncated in the middle of input file path. 

In my use case, the build system (Visual Studio msbuild) has many targets with dependencies between them. The files to be signed are managed in a collection and passed to xmlsec command target. Currently the build system understands that each invocation is independent of other invocations of xmlsec. It has its own way of launching them in parallel. I am not sure if I should try to manage parallel invocation myself outside of msbuild.
The PR you have seem the right approach. I will share my results once I can get my hands on new xmlsec tool.

Thanks a lot for fast turnaround time.

**Comment by lsh123** [comment]
"I want to give this change a try. How do I get access to prebuilt binaries with these changes?"

Your best bet is to build xmlsec yourself:

https://github.com/lsh123/xmlsec/tree/master/win32

You can find an example build here:

https://github.com/lsh123/xmlsec/blob/ad264336efc0370cc6afc1c39d99be17ac68d0de/.github/workflows/make-check.yml#L182

**Comment by videoguy** [comment]
One of my team members built xmlsec more than 6 yrs back and that is what we have been using. That team member is no longer with the company. It has been awhile since I built cygwin/mingw tools targetting Windows.

This is what I was able to do so far.
1) Downloaded libxml, libxslt and openssl and built them following the instructions under Win32 dir of xmlsec.
2) When I ran nmake on xmlsec, it is failing saying it is not able to find iconv.lib. I have mingw based Bash shell that I was using. This is something that got installed as part of Windows Git client. Then I started the search for iconv.lib package. I didn't make much progress.

Our company IT has blocked cygwin site. I couldn't install cygwin packages to start with. I am stuck there.

**Comment by lsh123** [comment]
Please take a look at:

https://github.com/lsh123/xmlsec/blob/ad264336efc0370cc6afc1c39d99be17ac68d0de/.github/workflows/make-check.yml#L182

The build script is run from MSVC shell. 

Re iconv: you can simply disable it with  configure.js iconv=no

**Comment by lsh123** [comment]
No updates for 2 weeks, closing

**Comment by videoguy** [comment]
We found different way to manage this. We found more than 90% of the files are duplicates of 10% of the files. We have a custom build job that locates the 10% and replace the 90% with symlinks to the 10%. That reduced the build time significantly. Signing 10% with xmlsec seemed to be fine for now.

But we do recognize that the version of xmlsec we use is pretty old. We have an action item to upgrade it to your stable release build.

**Comment by lsh123** [comment]
thanks for update!
